packages/itwin/tree-widget/src/test/trees/stateless/models-tree/internal/ModelsTreeVisibilityHandler.test.ts

```diff
@@ -0,0 +1,2266 @@
+                ])
+                .onSecondCall()
+                .returns([
+                  ...neverDrawnElements.map((elementId) => ({ elementId, modelId, categoryId })),
+                  { elementId: otherNeverDrawnElement, modelId: otherModelId, categoryId: otherCategoryId },
+                ]),
+            });
+            const idsCache = createFakeIdsCache({
+              modelCategories: new Map([
+                [modelId, [categoryId]],
+                [otherModelId, [otherCategoryId]],
+              ]),
+              categoryElements: new Map([
+                [categoryId, [...alwaysDrawnElements, ...neverDrawnElements]],
+                [otherCategoryId, [otherAlwaysDrawnElement, otherNeverDrawnElement]],
+              ]),
+            });
+            const { handler } = createHandler({ viewport, idsCache });
+            await handler.changeVisibility(node, visibility);
+            expect(viewport.alwaysDrawn).to.deep.eq(new Set([otherAlwaysDrawnElement]));
+            expect(viewport.neverDrawn).to.deep.eq(new Set([otherNeverDrawnElement]));
+          });
+        }
+
+        describe("on", () => {
+          it("marks itself visible", async () => {
+            const modelId = "0x1";
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({ viewport });
+            await handler.changeVisibility(node, true);
+            expect(viewport.addViewedModels).to.be.calledOnceWith(modelId);
+          });
+
+          testAlwaysAndNeverDrawnChildrenAccess(true);
+
+          it(`removes per model category overrides`, async () => {
+            const modelId = "0x1";
+            const categoryIds = ["0x2", "0x3", "0x4"];
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({
+              viewport,
+              idsCache: createFakeIdsCache({
+                modelCategories: new Map([[modelId, categoryIds]]),
+              }),
+            });
+            await handler.changeVisibility(node, true);
+
+            expect(viewport.perModelCategoryVisibility.clearOverrides).to.be.calledWith(modelId);
+          });
+        });
+
+        describe("off", () => {
+          it("marks itself hidden", async () => {
+            const modelId = "0x1";
+            const node = createModelHierarchyNode(modelId);
+            const viewport = createFakeSinonViewport();
+            const { handler } = createHandler({ viewport });
+            await handler.changeVisibility(node, false);
+            expect(viewport.changeModelDisplay).to.be.calledOnceWith(modelId, false);
+          });
+
+          testAlwaysAndNeverDrawnChildrenAccess(false);
+
+          it(`removes per model category overrides`, async () => {
```

Comment:
I think of the "model off" case we don't need to do anything besides removing it from model selector.

Reply:
Removed extra actions and unit tests